### DISCUSSION TITLE
Issue with table.view()

### DISCUSSION THREAD

**Discussion by jisnajc** [perspective-dev/perspective#2014] [Q&A]
I need to delete the perspective table in python as part of session clearing. But table.delete() function throws an error like this. 'Cannot delete a Table with active views still linked to it - call delete() on each view, and try again.'  When I tried to list the views associated to table with table.view() function , its not giving the views but views are listed in table._views array. How can I get the views of a table and delete it one by one before the table.delete?
Please find the below image for the error.
<img width="726" alt="Capture" src="https://user-images.githubusercontent.com/68993854/201392927-0440b502-f327-42bf-8348-ca7cfa9fb2c2.PNG">

Also I want to know the best approach for session handling in perspective python API. How can we clear the tables stored in session on closing of web socket by each user? I want to delete tables in session for that particular user. Is it auto handled by perspective? If not Is there any user - table mapping info available? Please guide me with the best possible approach.

**Comment by jisnajc** [comment]
Could you please help me out with the above issue. I am stuck with this.

**Comment by texodus** [comment]
> When I tried to list the views associated to table with table.view() function

Please read the docs - this is not what the `view()` function does.  [here](https://perspective.finos.org/docs/obj/perspective-python#viewcolumnsnone-group_bynone-split_bynone-aggregatesnone-sortnone-filternone-expressionsnone).

> How can I get the views of a table and delete it one by one before the table.delete?

Only you can answer that - what did you do with the `Table` after you created it?  You need to delete all references you've created previously with `.view()`.

> Please find the below image for the error.

This is an image of you not using your debugger correctly - you have a `PerspectiveManager` configured to run in async mode, you've spawned this `Table` in a background thread, and you are invoking commands from the interpreter's main thread.  The docs [cover this in detail](https://perspective.finos.org/docs/python#async-mode).

> Also I want to know the best approach for session handling in perspective python API. How can we clear the tables stored in session on closing of web socket by each user? I want to delete tables in session for that particular user.

The best approach is to not store tables per session, instead use perspective wasm and serialize the dataset to the client (use a fire-and-forget perspective-to-arrow conversion if you want).

What you're actually doing can still be managed quite easily by deleting the `Table`s when the session is deleted - you just need to delete any views you've created from this table, as per above (and [docs](https://perspective.finos.org/docs/view)).
